### PR TITLE
home-assistant-custom-lovelace-modules.zigbee2mqtt-networkmap: init at unstable-2023-12-06

### PR DESCRIPTION
**Pull request by azuwis**
## Description of changes

Home Assistant Custom Card to show Zigbee2mqtt network map, homepage: https://github.com/azuwis/zigbee2mqtt-networkmap

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/servers/home-assistant/custom-lovelace-modules/default.nix [hunk 1]
```diff
@@ -7,4 +7,6 @@
   mini-graph-card = callPackage ./mini-graph-card {};
 
   mini-media-player = callPackage ./mini-media-player {};
+
+  zigbee2mqtt-networkmap = callPackage ./zigbee2mqtt-networkmapmini-media-player {};
```
Comment (mweinelt) [inline pkgs/servers/home-assistant/custom-lovelace-modules/default.nix:11]:
```suggestion
  zigbee2mqtt-networkmap = callPackage ./zigbee2mqtt-networkmap {};
```

**Comment by azuwis** [comment]
@ofborg build home-assistant-custom-lovelace-modules.zigbee2mqtt-networkmap

**Comment by mweinelt** [comment]
I think having the yarn fetcher depend on `src` makes it IFD.

```
$ nix-instantiate -A home-assistant-custom-lovelace-modules.zigbee2mqtt-networkmap --option allow-import-from-derivation false
error:
       … while evaluating the 'name' attribute of a derivation

         at «none»:0: (source not available)

       … while evaluating a branch condition

         at /home/hexa/git/nixos/master/lib/strings.nix:1173:3:

         1172|   # First detect the common case of already valid strings, to speed those up
         1173|   if stringLength string <= 207 && okRegex string != null
             |   ^
         1174|   then unsafeDiscardStringContext string

       (stack trace truncated; use '--show-trace' to show the full trace)

       error: cannot build '/nix/store/p1d8z3rsp3mmywlrlddm0lsdq64xnrsv-source.drv^out' during evaluation because the option 'allow-import-from-derivation' is disabled

**Comment by mweinelt** [comment]
Nope, according to the eval the problem is the `package.json`

```

       … while realising the context of path '/nix/store/2mk2q4p45ffv13raqh77fhr2syjq5lp9-source/package.json'

       at /var/lib/ofborg/checkout/repo/38dca4e3aa6bca43ea96d2fcc04e8229/mr-est/ofborg-evaluator-7/lib/trivial.nix:313:24:

          312|   importJSON = path:
          313|     builtins.fromJSON (builtins.readFile path);
             |                        ^
          314|

       error: path '/nix/store/p1d8z3rsp3mmywlrlddm0lsdq64xnrsv-source.drv' is not valid
```

I think we need to import the `package.json` into nixpkgs and set `packageJSON` on the main derivation.

File: pkgs/servers/home-assistant/custom-lovelace-modules/zigbee2mqtt-networkmap/default.nix [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+  };
+
+  configurePhase = ''
+    cp -r $node_modules node_modules
+    chmod +w node_modules
+  '';
+
+  buildPhase = ''
+    yarn --offline build
+  '';
```
Comment (mweinelt) [inline pkgs/servers/home-assistant/custom-lovelace-modules/zigbee2mqtt-networkmap/default.nix:32]: Missing pre and post hooks

File: pkgs/servers/home-assistant/custom-lovelace-modules/zigbee2mqtt-networkmap/default.nix [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+    mkdir $out
+    cp -v dist/zigbee2mqtt-networkmap.js $out/
+
+    runHook postInstall
+  '';
+
+  dontFixup = true;
+
+  distPhase = "true";
```
Comment (mweinelt) [inline pkgs/servers/home-assistant/custom-lovelace-modules/zigbee2mqtt-networkmap/default.nix:45]:
```suggestion
  doDist = false;
```

**Review by mweinelt — CHANGES_REQUESTED** [review]

**Review by mweinelt — APPROVED** [review]

**Pull request merged**